### PR TITLE
added api interface to attach files to email

### PR DESCRIPTION
**Pull request by no1seman**
subj.

full description of api achanges in README.md and in init.lua comments
Closing: https://github.com/tarantool/smtp/issues/11

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -127,22 +127,58 @@ Example: `"Test Message"`.
...
 * `username` (string) -- a username for server authorization
 * `password` (string) -- a password for server authorization
+* `attachments` (table) -- a table (array) with attachments data
+  * `body` (any) attachment body contents
+  * `content_type` (string) -- set a content type (part of a Content-Type header,
+  defaults to 'text/plain')
+  * `charset` (string) -- set a charset (part of a Content-Type header, defaults
+  to 'UTF-8')
+  * `filename` (string) -- a string with filename will be shown in e-mail
+  * `base64_encode` (boolean) -- a boolean to base64 encode attachment content or not
```
Comment (Totktonada) [inline README.md:137]: What is default value?

File: README.md [hunk 2]
```diff
@@ -127,22 +127,58 @@ Example: `"Test Message"`.
...
+          content_type = 'application/json',
+          charset = 'UTF-8',
+          filename = 'json1.json',
+          base64_encode = true
+      },
+      {
+          body = json.encode('{"key2":"value2"}'),
+          content_type = 'application/json',
+          filename = 'json2.json',
+          base64_encode = false
```
Comment (Totktonada) [inline README.md:176]: Non-encoded JSON in the strange example. I guess the `base64_encode = false` variant is useful, when you already have a base64 encoded data.

File: README.md [hunk 3]
```diff
@@ -293,13 +329,13 @@ is done with a local test server, with none of the usual authorization options.)
...
 
-```
+```lua
 mails:get()
 ```
 
 And now the response should look like this:
 
-```
+```tarantool
```
Comment (Totktonada) [inline README.md:338]: Nit: There is no 'tarantool' language support on GitHub. I would use 'lua' or 'yaml'.

File: smtp/init.lua [hunk 4]
```diff
@@ -55,13 +55,13 @@ local smtp_new = function(opts)
...
-local function check_args(self, method)
-    if type(self) ~= 'table' then
-        error(check_args_fmt:format(method, method), 2)
-    end
-end
+-- local function check_args(self, method)
+--     if type(self) ~= 'table' then
+--         error(check_args_fmt:format(method, method), 2)
+--     end
+-- end
```
Comment (Totktonada) [inline smtp/init.lua:64]: It is better to remove the unused code: it's in the git history if anyone will need it (the same for `smtp_default` below). And It is better to move the dead code clean up into its own commit: this way it is easier to look over the history of changes.

File: smtp/init.lua [hunk 5]
```diff
@@ -39,7 +39,7 @@ local curl_mt
 --
 --  Parameters:
 --
---  max_connectionss -  Maximum number of entries in the connection cache */
+--  max_connections -  Maximum number of entries in the connection cache */
```
Comment (Totktonada) [inline smtp/init.lua:42]: I would move all those comment tweaks to its own commit (except ones that describe the functionality you added). It would be easier to look read the main changes this way.

File: smtp/init.lua [hunk 6]
```diff
@@ -167,15 +181,60 @@ curl_mt = {
                 header = header .. table.concat(opts.headers, '\r\n') .. '\r\n'
             end
 
-            local content_type = opts.content_type or 'text/plain'
-            local charset = opts.charset or 'UTF-8'
-            header = header .. 'Content-Type: ' .. content_type .. '; charset=' .. charset .. '\r\n'
+            local attachments = ''
```
Comment (Totktonada) [inline smtp/init.lua:184]: Nit: It is used only inside the `else` branch below.

File: smtp/init.lua [hunk 7]
```diff
@@ -167,15 +181,60 @@ curl_mt = {
...
+                                  content_disposition ..
+                                  attachment_body
+                end
+
+                local content_type = 'Content-Type: ' .. (opts.content_type or 'text/plain') ..
+                                         '; charset=' .. (opts.charset or  'UTF-8') ..';\r\n\r\n'
+
+                if opts.subject then
+                    header = header .. 'Subject: ' .. opts.subject .. '\r\n' .. ''
+                end
```
Comment (Totktonada) [inline smtp/init.lua:227]: I would perform an attempt to reduce code duplication where possible: either by changing of the code flow or by moving a common functionaliy into functions. Say, a function like `encode_content_type()` may be used as for the message itself as well as for attachments.

File: test/test.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#/bin/bash
```
Comment (Totktonada) [inline test/test.sh:1]: 1. Nit: There is not bash-specific here (any bourne shell compatible interpreter may run this code), so I would use `#!/bin/sh`. And, yep, the exclamation mark was forgotten.
2. I usually name such files with `.example` postfix and add a name without `.example` to gitignore. So a user may keep the repository clean and have secrets in a gitignored file. See token.txt.example [here](https://github.com/Totktonada/closed_issues) for example.
3. I would not name it 'test', because it does not verify correctness of the implementation. (The same for attach.test.lua.) I'll propose a name if something good will come into my mind.

File: smtp/init.lua [hunk 9]
```diff
@@ -167,15 +181,60 @@ curl_mt = {
...
+                local charset = opts.charset or 'UTF-8'
+                header = header .. 'Content-Type: ' .. content_type .. '; charset=' .. charset .. '\r\n'
+
+                if opts.subject then
+                    header = header .. 'Subject: ' .. opts.subject .. '\r\n'
+                end
+
+                body = header .. '\r\n' .. body
+            else
+                local MULTIPART_CONTENT_TYPE = 'Content-Type: multipart/mixed; boundary=MULTIPART-MIXED-BOUNDARY;\r\n'
```
Comment (Totktonada) [inline smtp/init.lua:196]: It would be good to mention exact standard you follow (at least because they're renewed from time to time).

**Comment by Totktonada** [comment]
> adeded

Typo (in the commit message). I would also use imperative mood for the commit message header ('add') and mention the issue in the commit message body. See more recommendations [here](https://www.tarantool.io/en/doc/latest/dev_guide/developer_guidelines/#how-to-write-a-commit-message).

File: README.md (hunk 2 above)
Comment (no1seman) [reply README.md:176]: OK, let's base64_encode will be true by default. I think base64_encode == false will be useful when need to send many tiny plain/text attachments. I think that option will be available.

File: .luacheckrc [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        'string',
+        'table',
+        'tonumber',
+        'tonumber64',
+        'tostring',
+        'type',
+        'unpack',
+        'xpcall',
+    },
+}
```
Comment (olegrok) [inline .luacheckrc:42]: There is no empty line at the end of file. Also seems you could use reduced version of .luacheckrc (see https://github.com/tarantool/cartridge/blob/master/.luacheckrc)

File: README.md (hunk 1 above)
Comment (Totktonada) [reply README.md:137]: Resolved in PR #23.

File: README.md (hunk 2 above)
Comment (Totktonada) [reply README.md:176]: Okay.

File: README.md (hunk 3 above)
Comment (Totktonada) [reply README.md:338]: Resolved in PR #22.

File: smtp/init.lua (hunk 4 above)
Comment (Totktonada) [reply smtp/init.lua:64]: Partially resolved in PR #23, but it's up to Yaroslav.

File: smtp/init.lua (hunk 5 above)
Comment (Totktonada) [reply smtp/init.lua:42]: Resolved in PR #21.

File: smtp/init.lua (hunk 6 above)
Comment (Totktonada) [reply smtp/init.lua:184]: Resolved in PR #23.

File: smtp/init.lua (hunk 7 above)
Comment (Totktonada) [reply smtp/init.lua:227]: Mostly resolved in PR #23.

File: test/test.sh (hunk 8 above)
Comment (Totktonada) [reply test/test.sh:1]: The file was replaced by the actual test in PR #23, so the comment is not actual anymore.

File: smtp/init.lua (hunk 9 above)
Comment (Totktonada) [reply smtp/init.lua:196]: Resolved in PR #23.

File: .luacheckrc (hunk 10 above)
Comment (Totktonada) [reply .luacheckrc:42]: The file was excluded from PR #23, so the coment becomes not actual.
